### PR TITLE
Use scheme/host from request for token redirect

### PR DESCRIPTION
**Pull request by liggitt**
Fixes https://github.com/openshift/origin/issues/10193 and https://bugzilla.redhat.com/show_bug.cgi?id=1367610

### CODE REVIEW

File: pkg/util/httprequest/httprequest.go [hunk 1]
```diff
@@ -38,3 +39,50 @@ func PrefersHTML(req *http.Request) bool {
 
 	return false
 }
+
+func SchemeHost(req *http.Request) (scheme, host string) {
```
Comment (smarterclayton) [inline pkg/util/httprequest/httprequest.go]: Surprised something like this doesn't already exist.  Needs Godoc

Comment (liggitt) [inline pkg/util/httprequest/httprequest.go]: doc and tests forthcoming. WIP :)

**Comment by liggitt** [comment]
[test]

**Comment by liggitt** [comment]
@aweiteka @eformat fyi

**Comment by liggitt** [comment]
@miminar @mfojtik PTAL

File: pkg/dockerregistry/server/auth.go [hunk 2]
```diff
@@ -147,6 +151,36 @@ var (
...
+	}
+	if len(tokenRealm.RawQuery) > 0 || len(tokenRealm.Fragment) > 0 {
+		return nil, fmt.Errorf("%s config option may not contain query parameters or a fragment", TokenRealmKey)
+	}
+	if len(tokenRealm.Path) > 0 {
+		return nil, fmt.Errorf("%s config option may not contain a path (%q was specified)", TokenRealmKey, tokenRealm.Path)
+	}
+
+	// pin to "/openshift/token"
+	tokenRealm.Path = defaultTokenPath
```
Comment (miminar) [inline pkg/dockerregistry/server/auth.go]: Why calling it *default* when it cannot be overridden?

File: pkg/util/httprequest/httprequest.go [hunk 3]
```diff
@@ -38,3 +39,64 @@ func PrefersHTML(req *http.Request) bool {
...
+
+	scheme := ""
+	switch {
+	case len(forwardedProto) > 0:
+		scheme = forwardedProto
+	case req.TLS != nil:
+		scheme = "https"
+	case len(req.URL.Scheme) > 0:
+		scheme = req.URL.Scheme
+	default:
```
Comment (miminar) [inline pkg/util/httprequest/httprequest.go]: Default to `https` if given port 443?

File: pkg/dockerregistry/server/auth.go (hunk 2 above)
Comment (liggitt) [inline pkg/dockerregistry/server/auth.go]: Started out being able to be overridden, but I couldn't come up with a reasonable way to know what paths to disallow.

File: pkg/cmd/dockerregistry/dockerregistry.go [hunk 4]
```diff
@@ -82,8 +59,16 @@ func Execute(configFile io.Reader) {
...
-	if usingOpenShiftAuth {
-		app.NewRoute().Methods("GET").PathPrefix(tokenPath).Handler(server.NewTokenHandler(ctx, server.DefaultRegistryClient))
+	if options, usingOpenShiftAuth := config.Auth[server.OpenShiftAuth]; usingOpenShiftAuth {
+		tokenRealm, err := server.TokenRealm(options)
+		if err != nil {
+			log.Fatalf("Error setting up token auth: %s", err)
+		}
+		err = app.NewRoute().Methods("GET").PathPrefix(tokenRealm.Path).Handler(server.NewTokenHandler(ctx, server.DefaultRegistryClient)).GetError()
+		if err != nil {
+			log.Fatalf("Error setting up token auth: %s", err)
```
Comment (miminar) [inline pkg/cmd/dockerregistry/dockerregistry.go]: nit: `"Error setting up token endpoint at %s: %v", tokenRealm.Path, err`

Comment (liggitt) [inline pkg/cmd/dockerregistry/dockerregistry.go]: updated

File: pkg/util/httprequest/httprequest.go (hunk 3 above)
Comment (liggitt) [inline pkg/util/httprequest/httprequest.go]: updated

**Comment by liggitt** [comment]
[merge]

**Pull request merged**